File: src/lib/charts/adapters/uk.ts
import { Chart, ChartCategory, ChartProvider } from '../types';
import * as cheerio from 'cheerio';

export class UKAdapter implements ChartProvider {
  name = 'UK';
  
  private baseUrl: string;
  private pageUrl: string = '';

  constructor() {
    // Defines the base URL for the AIRAC cycle
    // We use a dedicated ENV variable for NATS as format differs or cycle might not align with SIA
    const airacPart = process.env.NEXT_PUBLIC_NATS_AIRAC_URL_PART || '2026-01-22-AIRAC'; 
    this.baseUrl = `https://www.aurora.nats.co.uk/htmlAIP/Publications/${airacPart}/html/eAIP/`;
  }

  async getCharts(icao: string): Promise<Chart[]> {
    // UK eAIP format: EG-AD-2.EGLL-en-GB.html
    this.pageUrl = `${this.baseUrl}EG-AD-2.${icao}-en-GB.html`;
    
    try {
      const response = await fetch(this.pageUrl);
      if (!response.ok) {
        if (response.status === 404) return [];
        throw new Error(`UK AIP fetch failed: ${response.statusText}`);
      }

      const html = await response.text();
      return this.parseHtml(html);
    } catch (error) {
       console.error('UK Adapter Error:', error);
       throw error;
    }
  }

  private parseHtml(html: string): Chart[] {
    const $ = cheerio.load(html);
    const charts: Chart[] = [];

    // UK AIP usually lists charts in tables (AD 2.24)
    // We look for all PDF links in the document to coverage embedded charts too
    $('a[href$=".pdf"]').each((_, element) => {
      const $el = $(element);
      const href = $el.attr('href');
      if (!href) return;

      const title = $el.text().trim() || $el.attr('title') || '';
      
      // Attempt to get more context if the link text is short (e.g., just "PDF")
      let fullTitle = title;
      const row = $el.closest('tr');
      const prevRow = row.prev();

      // NATS layout often puts the title in the previous row
      // Row 1: <p>AERODROME CHART - ICAO</p>
      // Row 2: <a href...>AD 2.EGLL...</a>
      if (prevRow.length > 0) {
          const prevText = prevRow.text().replace(/\s+/g, ' ').trim();
          // If previous text exists and seems descriptive (longer than 3 chars), prefer it or combine it
          if (prevText.length > 3) {
              fullTitle = prevText;
          }
      } else if (row.length > 0) {
          // Fallback: If inside a table but no prev row (unlikely for this layout), check current row text
          // But exclude the link text itself to avoid duplication
          const rowText = row.text().replace(title, '').replace(/\s+/g, ' ').trim();
          if (rowText.length > 3) fullTitle = rowText;
      }
      
      const decodedFilename = decodeURIComponent(href.split('/').pop() || '');
      const typeInfo = this.identifyType(fullTitle, decodedFilename);
      
      // If we can't identify the type, it might not be a chart (e.g. textual annex)
      // UK AIP charts usually have clear names "Instrument Approach Chart", "Aerodrome Chart"
      if (!typeInfo) return;

      const absoluteUrl = new URL(href, this.pageUrl).toString();
      const tags = this.generateTags(fullTitle);
      const { subtitle, page } = this.extractSubtitle(fullTitle, typeInfo.category);

      charts.push({
        id: absoluteUrl,
        source: 'UK',
        category: typeInfo.category,
        subtitle: subtitle || title, // Fallback to link text
        filename: decodedFilename,
        url: absoluteUrl,
        page,
        tags
      });
    });

    return this.postProcess(charts);
  }

  private identifyType(text: string, filename: string): { category: ChartCategory } | null {
    const t = text.toUpperCase();
    const f = filename.toUpperCase();

    if (t.includes('AERODROME CHART') || t.includes('AERODROME OBSTACLE')) return { category: ChartCategory.AERODROME };
    if (t.includes('AIRCRAFT PARKING') || t.includes('DOCKING')) return { category: ChartCategory.PARKING };
    if (t.includes('GROUND MOVEMENT') || t.includes('TAXI')) return { category: ChartCategory.GROUND };
    
    // Approaches
    if (t.includes('INSTRUMENT APPROACH') || t.includes('IAC') || f.includes('IAC')) return { category: ChartCategory.IAC };
    if (t.includes('VISUAL APPROACH') || t.includes('VAC')) return { category: ChartCategory.VAC };
    
    // Departments / Arrivals
    if (t.includes('STANDARD DEPARTURE') || t.includes('SID')) return { category: ChartCategory.SID };
    if (t.includes('STANDARD ARRIVAL') || t.includes('STAR')) return { category: ChartCategory.STAR };
    
    // UK specific "Initial Approach Procedures"
    if (t.includes('INITIAL APPROACH')) return { category: ChartCategory.IAC }; 

    if (t.includes('ATC SURVEILLANCE') || t.includes('MINIMUM ALTITUDE')) return { category: ChartCategory.IAC };

    return null;
  }

  private generateTags(text: string): string[] {
    const tags: string[] = [];
    const t = text.toUpperCase();

    // Runways
    // Look for RWY 09L, RWY 27, Runway 27R etc.
    const rwyMatch = t.match(/(?:RWY|RUNWAY)[\s]*(\d{2}[LRC]?)/g);
    if (rwyMatch) {
       rwyMatch.forEach(m => {
           const r = m.replace(/(RWY|RUNWAY)[\s]*/, '').trim();
           tags.push(r);
       });
    }

    // Procedures
    if (t.includes('ILS')) tags.push('ILS');
    if (t.includes('LOC')) tags.push('LOC');
    if (t.includes('RNP')) tags.push('RNP');
    if (t.includes('RNAV')) tags.push('RNAV');
    if (t.includes('VOR')) tags.push('VOR');
    if (t.includes('NDB')) tags.push('NDB');
    if (t.includes('DME')) tags.push('DME');
    if (t.includes('VISUAL')) tags.push('VISUAL');

    // CAT categories (UK often lists CAT II/III in title)
    if (t.includes('CAT II') || t.includes('CAT III')) tags.push('ILS I/II/III');

    return Array.from(new Set(tags));
  }

  private extractSubtitle(fullTitle: string, category: ChartCategory): { subtitle: string, page?: string } {
    // UK Titles are often verbose: "AD 2.24-1 Aerodrome Chart - ICAO - 1"
    
    // Clean up common prefixes/suffixes
    let subtitle = fullTitle
        .replace(/^[A-Z]{2}\s+AD\s+2\.[^ ]+\s+/, '') // Remove "EG AD 2.EGLL..."
        .replace(/AD 2\.[^ ]+\s+/, '')
        // Specific UK cleanups requested by user
        .replace(/^INSTRUMENT APPROACH CHART\s*/i, '')
        .replace(/INITIAL APPROACH PROCEDURES?\s*/i, 'INITIAL ')
        .replace(/STANDARD ARRIVAL CHART\s*(?:-\s*INSTRUMENT\s*)?(?:\(\s*\))?\s*/i, '')
        .replace(/STANDARD DEPARTURE CHART\s*(?:-\s*INSTRUMENT\s*)?(?:\(?SID\)?\s*)?(?:\(\s*\))?\s*/i, '')
        // General cleanup
        .replace(category.toString(), '') // Remove category name if present verbatim
        .replace('Chart', '')
        .replace(' - ICAO', '')
        .replace(/\s+/g, ' ')
        .trim();

    // Remove file extensions or parens
    subtitle = subtitle.replace(/\.pdf$/i, '');

    return { subtitle };
  }

  private postProcess(charts: Chart[]): Chart[] {
      // Deduplicate by URL
      const unique = new Map<string, Chart>();
      charts.forEach(c => unique.set(c.url, c));
      return Array.from(unique.values());
  }
}

File: crates/atc-book/src/adapters/uk.rs
use regex::Regex;
use scraper::{Html, Selector};

use chrono::Duration;

use crate::airac::AiracCycle;
use crate::models::{Chart, ChartCategory, ChartSource};

/// Fetch charts from the UK NATS eAIP (Aurora) for a given ICAO code.
pub async fn fetch_charts(icao: &str, airac: &AiracCycle) -> Result<Vec<Chart>, String> {
    let client = reqwest::Client::builder()
        .build()
        .map_err(|e| format!("HTTP client error: {e}"))?;
    let mut tried_urls = Vec::new();

    for (part, code) in candidate_airac_parts(airac) {
        let url = build_uk_url(&part, icao);
        tried_urls.push(url.clone());

        let resp = client
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("UK NATS request failed: {e}"))?;

        if resp.status().is_success() {
            let html = resp.text().await.map_err(|e| format!("Read error: {e}"))?;
            return parse_uk_html(&html, icao, &code);
        }

        if resp.status().as_u16() != 404 {
            return Err(format!("HTTP {}", resp.status()));
        }
    }

    Err(format!("HTTP 404 (tried {})", tried_urls.join(", ")))
}

fn build_uk_url(airac_part: &str, icao: &str) -> String {
    format!(
        "https://www.aurora.nats.co.uk/htmlAIP/Publications/{}/html/eAIP/EG-AD-2.{}-en-GB.html",
        airac_part,
        icao.to_uppercase()
    )
}

fn candidate_airac_parts(airac: &AiracCycle) -> Vec<(String, String)> {
    let mut parts = Vec::new();
    let mut date = airac.start_date;
    for _ in 0..3 {
        let cycle = AiracCycle::for_date(date);
        parts.push((cycle.nats_airac_part(), cycle.code));
        date = date - Duration::days(28);
    }
    parts
}

fn parse_uk_html(html: &str, icao: &str, airac_code: &str) -> Result<Vec<Chart>, String> {
    let doc = Html::parse_document(html);
    let a_sel = Selector::parse("a[href]").unwrap();
    let rwy_re = Regex::new(r"(?i)(?:RWY|RUNWAY)\s*(\d{2}[LRC]?)").unwrap();

    let mut charts = Vec::new();

    for (i, elem) in doc.select(&a_sel).enumerate() {
        let href = match elem.value().attr("href") {
            Some(h) if h.to_lowercase().ends_with(".pdf") => h,
            _ => continue,
        };

        let link_text = elem.text().collect::<Vec<_>>().join(" ");
        let upper = link_text.to_uppercase();

        let category = detect_uk_category(&upper);
        let tags = extract_uk_tags(&upper, &rwy_re);
        let runways = extract_uk_runways(&upper, &rwy_re);

        let provider_relative_url = if href.starts_with("http") {
            href.to_string()
        } else {
            href.trim_start_matches("./").to_string()
        };

        let filename = href.rsplit('/').next().unwrap_or(href).to_string();
        let subtitle = link_text.trim().to_string();
        let id = format!("{}-UK-{:02}", icao.to_uppercase(), i);

        charts.push(Chart {
            id,
            source: ChartSource::Uk,
            category,
            subtitle,
            filename,
            provider_relative_url,
            linked_provider_relative_urls: Vec::new(),
            airac_code: airac_code.to_string(),
            page: None,
            tags,
            runways,
            custom_title: None,
        });
    }

    Ok(charts)
}

fn detect_uk_category(upper: &str) -> ChartCategory {
    if upper.contains("AERODROME CHART") || upper.contains("ADC") {
        ChartCategory::Aerodrome
    } else if upper.contains("AIRCRAFT PARKING") || upper.contains("PARKING") {
        ChartCategory::Parking
    } else if upper.contains("GROUND MOVEMENT") || upper.contains("GMC") {
        ChartCategory::Ground
    } else if upper.contains("STANDARD DEPARTURE") || upper.contains("SID") {
        ChartCategory::Sid
    } else if upper.contains("STANDARD ARRIVAL") || upper.contains("STAR") {
        ChartCategory::Star
    } else if upper.contains("INSTRUMENT APPROACH") || upper.contains("IAC") {
        ChartCategory::Iac
    } else if upper.contains("VISUAL APPROACH") {
        ChartCategory::Vac
    } else {
        ChartCategory::Other
    }
}

fn extract_uk_tags(upper: &str, rwy_re: &Regex) -> Vec<String> {
    let mut tags = Vec::new();

    if upper.contains("ILS") {
        tags.push("ILS".to_string());
    }
    if upper.contains("LOC") {
        tags.push("LOC".to_string());
    }
    if upper.contains("RNP") {
        tags.push("RNP".to_string());
    }
    if upper.contains("RNAV") {
        tags.push("RNAV".to_string());
    }
    if upper.contains("VOR") {
        tags.push("VOR".to_string());
    }
    if upper.contains("NDB") {
        tags.push("NDB".to_string());
    }
    if upper.contains("DME") {
        tags.push("DME".to_string());
    }
    if upper.contains("VISUAL") {
        tags.push("VISUAL".to_string());
    }
    if upper.contains("CAT II") || upper.contains("CAT III") {
        if upper.contains("CAT III") {
            tags.push("CAT III".to_string());
        } else {
            tags.push("CAT II".to_string());
        }
    }

    for cap in rwy_re.captures_iter(upper) {
        tags.push(cap[1].to_string());
    }

    tags.sort();
    tags.dedup();
    tags
}

fn extract_uk_runways(upper: &str, rwy_re: &Regex) -> Vec<String> {
    let mut rwys: Vec<String> = rwy_re
        .captures_iter(upper)
        .map(|c| c[1].to_string())
        .collect();
    rwys.sort();
    rwys.dedup();
    rwys
}
